### PR TITLE
feat: type characteristic for nonempty types

### PR DESCRIPTION
**Pull request by RustanLeino**
This PR introduces the type characteristic `(00)`, which says a type is nonempty. The syntax is less than perfect, but it can be read as "a little weaker than `(0)`", where `(0)` is the type characteristic that says a type is not only nonempty but also has a compilable value.

Other details of this PR:

- chore: Rename `WKind.None` to `WKind.CompiledZero`
- fix: Print `ghost` keyword for all let(-such-that) expressions where it's been given syntactically
- Mark variables of ghost-context let(-such-that) expressions ghost
- fix: Pass correct ghost context into CheckExpression
- fix: Call CheckExpression on witness expressions
- refactor: Rename values of type-characteristic enum
- fix: Typo in comment, pointed out in previous PR comments

### CODE REVIEW

File: Source/Dafny/DafnyAst.cs [hunk 1]
```diff
@@ -4846,7 +4873,7 @@ public interface ICallable : ICodeContext
     /// </summary>
     bool InferredDecreases { get; set; }
   }
-
+
```
Comment (davidcok) [inline Source/Dafny/DafnyAst.cs:4876]: No trailing white space

File: Test/cloudmake/CloudMake-ParallelBuilds.dfy [hunk 2]
```diff
@@ -150,10 +150,10 @@ abstract module M0 {
 
   datatype Reason = rCompatibility | rValidity
 
-  type Path(==,0)  // TODO: this is a good candidate for a "ghost (0)" marking (meaning nonempty type)
+  type Path(==,00)
   function Loc(cmd: string, deps: set<Path>, exp: string): Path
 
-  type Artifact(0)  // TODO: this is a good candidate for a "ghost (0)" marking (meaning nonempty type)
+  type Artifact(0)
```
Comment (davidcok) [inline Test/cloudmake/CloudMake-ParallelBuilds.dfy:156]: Given the comment, I would have expected 00 here

File: Test/server/counterexample_none.transcript.expect [hunk 3]
```diff
@@ -2,6 +2,6 @@
 
 Verifying Impl$$_module.__default.AbsCorrect ...
   [1 proof obligation]  verified
-COUNTEREXAMPLE_START {"States":[]} COUNTEREXAMPLE_END
+COUNTEREXAMPLE_START {"States":[{"Column":0,"Line":0,"Name":"<initial>","Variables":[]}]} COUNTEREXAMPLE_END
```
Comment (davidcok) [inline Test/server/counterexample_none.transcript.expect:5]: I think we had to revert this change in the past.

**Review by davidcok — CHANGES_REQUESTED** [review]
00 does not exactly connote non-empty to me. Perhaps + (as in one or more) or !{} or {+} or 1+. But just suggestions.

File: Test/cloudmake/CloudMake-ParallelBuilds.dfy (hunk 2 above)
Comment (RustanLeino) [reply Test/cloudmake/CloudMake-ParallelBuilds.dfy]: Done. (Due to an intermediate bug, I temporarily changed it back to `(0)` and then forgot to re-change to it `(00)`.)

File: Test/server/counterexample_none.transcript.expect (hunk 3 above)
Comment (RustanLeino) [reply Test/server/counterexample_none.transcript.expect]: Done
